### PR TITLE
Release Service: Add GRPC APIs

### PR DESCRIPTION
**Pull request by vmilovanovicc**
## Changes
Added the API implementation for Publish, Listrelease and GetLatestRelease.
Added the test cases for the same.

### CODE REVIEW

File: internal/release/services/release/release.go [hunk 1]
```diff
@@ -51,9 +68,174 @@ func (s *ReleaseService) RegisterWithServer(grpcServer grpc.ServiceRegistrar) er
...
+		return TimeA.Before(TimeB)
+	})
+
+	return releases
+}
+
+// TODO: OPTIMIZE!!!
+// GetLatestRelease retrieves the latest release.
+func (s *ReleaseService) GetLatestRelease(ctx context.Context, request *releaseSvc.ListReleasesRequest) (*releaseSvc.ListReleasesResponse, error) {
+	log.Printf("Getting latest published release.")
```
Comment (adcharre) [inline internal/release/services/release/release.go:197]: The optimisation here is to add another table which only stores the most recent release for a unique tuple of (type,org,name) and is updated whenever a new release is published.
For simplicity we could store the same data as we store in the "Release" log table

Comment (adcharre) [reply internal/release/services/release/release.go:197]: To be clear this doesn't need to be done in this story - we can add another US to do this optimisation later.

Comment (vmilovanovicc) [reply internal/release/services/release/release.go:197]: Marking it as resolved. We'll add a ticket to the backlog.

File: internal/release/services/release/release.go [hunk 2]
```diff
@@ -54,6 +71,165 @@ func (s *ReleaseService) RegisterWithServer(grpcServer grpc.ServiceRegistrar) er
...
+	}
+
+	log.Println("New release created.")
+	return ReleasePublished, nil
+}
+
+// ListReleases Retrieves all releases.
+// Only releases that have been published should be reported.
+func (s *ReleaseService) ListReleases(ctx context.Context, request *releaseSvc.ListReleasesRequest) (*releaseSvc.ListReleasesResponse, error) {
+
```
Comment (adcharre) [inline internal/release/services/release/release.go:120]: Lets add Otel trace here as well

File: internal/release/services/release/release.go [hunk 3]
```diff
@@ -54,6 +71,165 @@ func (s *ReleaseService) RegisterWithServer(grpcServer grpc.ServiceRegistrar) er
...
+		return TimeA.Before(TimeB)
+	})
+
+	return releases
+}
+
+// TODO: OPTIMIZE!!!
+// GetLatestRelease retrieves the latest release.
+func (s *ReleaseService) GetLatestRelease(ctx context.Context, request *releaseSvc.ListReleasesRequest) (*releaseSvc.ListReleasesResponse, error) {
+	log.Printf("Getting latest published release.")
```
Comment (adcharre) [inline internal/release/services/release/release.go:191]: Add otel trace and remove log statements.

File: internal/release/services/release/release_test.go [hunk 4]
```diff
@@ -66,3 +73,276 @@ func Test_RegisterReleaseWithServer(t *testing.T) {
...
+				{
+					Title: "test title",
+					Url:   "http://www.google.com",
+				},
+			},
+		}
+
+		res, err := svc.Publish(context.TODO(), req)
+
+		if err != nil {
```
Comment (adcharre) [inline internal/release/services/release/release_test.go:104]: (for info) We can simplify our unit tests by using the "require" package, see https://pkg.go.dev/github.com/stretchr/testify/require#NoError for an example.

File: internal/release/services/release/release.go [hunk 5]
```diff
@@ -54,6 +71,165 @@ func (s *ReleaseService) RegisterWithServer(grpcServer grpc.ServiceRegistrar) er
 	return nil
 }
 
+// Publishes a new release.
+func (s *ReleaseService) Publish(ctx context.Context, request *release.PublishRequest) (*release.PublishResponse, error) {
+
+	log.Println("Creating new release.")
```
[the remark this answers is not in the document]
Comment (aflahahamed) [reply internal/release/services/release/release.go]: Added the otel trace for the function and removed logs

File: internal/release/services/release/release.go (hunk 2 above)
Comment (aflahahamed) [reply internal/release/services/release/release.go:118]: Added the otel trace for the function and removed logs

File: internal/release/services/release/release.go (hunk 3 above)
Comment (aflahahamed) [reply internal/release/services/release/release.go]: Added the otel trace for the function and removed logs

File: internal/release/services/release/release.go [hunk 6]
```diff
@@ -54,6 +72,170 @@ func (s *ReleaseService) RegisterWithServer(grpcServer grpc.ServiceRegistrar) er
...
+	return ReleasePublished, nil
+}
+
+// ListReleases Retrieves all releases.
+// Only releases that have been published should be reported.
+func (s *ReleaseService) ListReleases(ctx context.Context, request *releaseSvc.ListReleasesRequest) (*releaseSvc.ListReleasesResponse, error) {
+
+	span := trace.SpanFromContext(ctx)
+	span.SetAttributes(
+		attribute.String("release.list", "Lists releases based on filters"),
```
Comment (adcharre) [inline internal/release/services/release/release.go:121]: Replace this with the values in the `request` argument, ie I would expect to see MaxAgeSeconds, types, organizations and page as attributes.

File: internal/release/services/release/release.go [hunk 7]
```diff
@@ -54,6 +72,170 @@ func (s *ReleaseService) RegisterWithServer(grpcServer grpc.ServiceRegistrar) er
...
+	return releases
+}
+
+// TODO: OPTIMIZE!!!
+// GetLatestRelease retrieves the latest release.
+func (s *ReleaseService) GetLatestRelease(ctx context.Context, request *releaseSvc.ListReleasesRequest) (*releaseSvc.ListReleasesResponse, error) {
+
+	span := trace.SpanFromContext(ctx)
+	span.SetAttributes(
+		attribute.String("release", "Lists the latest release"),
```
Comment (adcharre) [inline internal/release/services/release/release.go:197]: Same comment as for ListReleases() this should be something useful!

File: internal/release/services/release/release.go [hunk 8]
```diff
@@ -54,6 +73,184 @@ func (s *ReleaseService) RegisterWithServer(grpcServer grpc.ServiceRegistrar) er
...
+// Only releases that have been published should be reported.
+func (s *ReleaseService) ListReleases(ctx context.Context, request *releaseSvc.ListReleasesRequest) (*releaseSvc.ListReleasesResponse, error) {
+
+	span := trace.SpanFromContext(ctx)
+
+	// attribute does not support Uint64 so converting to int64
+	MaxAgeSeconds := convertUint64ToInt64(request.GetMaxAgeSeconds())
+
+	span.SetAttributes(
+		attribute.StringSlice("release.organization", request.GetOrganizations()),
```
Comment (adcharre) [inline internal/release/services/release/release.go:126]:
```suggestion
		attribute.StringSlice("release.organizations", request.GetOrganizations()),
```

File: internal/release/services/release/release.go [hunk 9]
```diff
@@ -54,6 +73,184 @@ func (s *ReleaseService) RegisterWithServer(grpcServer grpc.ServiceRegistrar) er
...
+// GetLatestRelease retrieves the latest release.
+func (s *ReleaseService) GetLatestRelease(ctx context.Context, request *releaseSvc.ListReleasesRequest) (*releaseSvc.ListReleasesResponse, error) {
+
+	span := trace.SpanFromContext(ctx)
+
+	// attribute does not support Uint64 so converting to int64
+	MaxAgeSeconds := convertUint64ToInt64(request.GetMaxAgeSeconds())
+
+	span.SetAttributes(
+		attribute.StringSlice("release.organization", request.GetOrganizations()),
```
Comment (adcharre) [inline internal/release/services/release/release.go:209]:
```suggestion
		attribute.StringSlice("release.organizations", request.GetOrganizations()),
```

File: internal/release/services/release/release.go [hunk 10]
```diff
@@ -54,6 +73,184 @@ func (s *ReleaseService) RegisterWithServer(grpcServer grpc.ServiceRegistrar) er
 	return nil
 }
 
+// Publishes a new release.
+func (s *ReleaseService) Publish(ctx context.Context, request *release.PublishRequest) (*release.PublishResponse, error) {
+
+	span := trace.SpanFromContext(ctx)
+	span.SetAttributes(
+		attribute.String("release.name", request.GetName()),
```
Comment (adcharre) [inline internal/release/services/release/release.go:81]: Lets add "release.type" and "release.organization" here. I don't think we need links or description.

File: internal/release/services/release/release.go [hunk 11]
```diff
@@ -54,6 +73,186 @@ func (s *ReleaseService) RegisterWithServer(grpcServer grpc.ServiceRegistrar) er
...
+		attribute.String("release.version", request.GetVersion()),
+		attribute.String("release.type", request.GetType()),
+		attribute.String("release.organization", request.GetOrganization()),
+	)
+
```
Comment (adcharre) [inline internal/release/services/release/release.go:86]: You are missing a `defer span.end()`

File: internal/release/services/release/release.go [hunk 12]
```diff
@@ -54,6 +73,186 @@ func (s *ReleaseService) RegisterWithServer(grpcServer grpc.ServiceRegistrar) er
...
+		Item:      av,
+		TableName: aws.String(ReleaseTableName),
+	}
+
+	if _, err = s.Db.PutItem(ctx, in); err != nil {
+		span.RecordError(err)
+		return nil, PublishReleaseError
+	}
+
+	log.Println("New release created.")
```
Comment (adcharre) [inline internal/release/services/release/release.go:114]: Remove

File: internal/release/services/release/release.go [hunk 13]
```diff
@@ -54,6 +73,186 @@ func (s *ReleaseService) RegisterWithServer(grpcServer grpc.ServiceRegistrar) er
...
+	return ReleasePublished, nil
+}
+
+// ListReleases Retrieves all releases.
+// Only releases that have been published should be reported.
+func (s *ReleaseService) ListReleases(ctx context.Context, request *releaseSvc.ListReleasesRequest) (*releaseSvc.ListReleasesResponse, error) {
+
+	span := trace.SpanFromContext(ctx)
+
```
Comment (adcharre) [inline internal/release/services/release/release.go:123]: Add a `defer span.end()`

File: internal/release/services/release/release.go [hunk 14]
```diff
@@ -54,6 +73,186 @@ func (s *ReleaseService) RegisterWithServer(grpcServer grpc.ServiceRegistrar) er
...
+		FilterExpression:          expr.Filter(),
+		TableName:                 aws.String(ReleaseTableName),
+	}
+
+	response, err := s.Db.Scan(ctx, scanQueryInputs)
+	if err != nil {
+		span.RecordError(err)
+		return nil, err
+	}
+	log.Println("Filtered Release Count: ", len(response.Items))
```
Comment (adcharre) [inline internal/release/services/release/release.go:161]: Add as an attribute to the span

File: internal/release/services/release/release.go [hunk 15]
```diff
@@ -54,6 +73,186 @@ func (s *ReleaseService) RegisterWithServer(grpcServer grpc.ServiceRegistrar) er
...
+
+	return releases
+}
+
+// TODO: OPTIMIZE!!!
+// GetLatestRelease retrieves the latest release.
+func (s *ReleaseService) GetLatestRelease(ctx context.Context, request *releaseSvc.ListReleasesRequest) (*releaseSvc.ListReleasesResponse, error) {
+
+	span := trace.SpanFromContext(ctx)
+
```
Comment (adcharre) [inline internal/release/services/release/release.go:206]: Add a `defer span.end()`

**Review by adcharre — COMMENTED** [review]
Almost there

File: internal/release/services/release/release.go [hunk 16]
```diff
@@ -54,6 +73,188 @@ func (s *ReleaseService) RegisterWithServer(grpcServer grpc.ServiceRegistrar) er
...
+	scanQueryInputs := &dynamodb.ScanInput{
+		TableName: aws.String(ReleaseTableName),
+	}
+
+	response, err := s.Db.Scan(ctx, scanQueryInputs)
+	if err != nil {
+		span.RecordError(err)
+		return nil, err
+	}
+	log.Println(response)
```
Comment (adcharre) [inline internal/release/services/release/release.go:228]: Missed this one on my last review - remove as I don't believe it's adding anything....

**Review by adcharre — APPROVED** [review]

**Pull request merged**